src/rard/research/migrations/0061_order_in_book.py

```diff
@@ -0,0 +1,196 @@
+                    link.save()
+
+        for link in worklinks:
+            order_in_book_for_link_class(link)
+
+    def reindex_work_by_book(apps, schema_editor):
+        """Update order of links with respect to work, taking into account book__order and order_in_book"""
+
+        worklinks = ["FragmentLink", "TestimoniumLink", "AppositumFragmentLink"]
+
+        Work = apps.get_model("research", "Work")
+
+        def reindex_fragment_and_testimonium_links(antiquarian):
+            to_reorder = [
+                antiquarian.fragmentlinks.all()
+                .order_by(
+                    "work__worklink__order",
+                    "work_order",
+                    "book__order",
+                    "order_in_book",
+                )
+                .distinct(),
+                # We want testimonium links without works to be ordered first
+                # but for some reason adding "-work__isnull" to the order_by
+                # breaks the rest of the ordering. Using itertools chain as
+                # a workaround does the job.
+                itertools.chain(
+                    antiquarian.testimoniumlinks.all()
+                    .filter(work__isnull=True)
+                    .distinct(),
+                    antiquarian.testimoniumlinks.all()
+                    .filter(work__isnull=False)
+                    .order_by(
+                        "work__worklink__order",
+                        "work_order",
+                        "book__order",
+                        "order_in_book",
+                    )
+                    .distinct(),
+                ),
+                antiquarian.appositumfragmentlinks.all()
+                .order_by(
+                    "work__worklink__order",
+                    "work_order",
+                    "book__order",
+                    "order_in_book",
+                )
+                .distinct(),
+            ]
+
+            for qs in to_reorder:
+                for count, link in enumerate(qs):
+                    if link.order != count:
+                        link.order = count
+                        link.save()
+
+        # parent function
+        for work in Work.objects.all():
+            for link_class in worklinks:
+                link = apps.get_model("research", link_class)
+
+                to_reorder = (
+                    link.objects.filter(work=work)
+                    .order_by("book__order", "order_in_book")
+                    .distinct()
+                )
+
+                for count, link in enumerate(to_reorder):
+                    if link.book.work.unknown == True:
+                        link.work_order = link.order_in_book
+                        link.save()
+                    else:
+                        if link.work_order != count:
+                            link.work_order = count
+                            link.save()
+                    reindex_fragment_and_testimonium_links(link.antiquarian)
```

Comment:
Calling reindex_fragment_and_testimonium_links within this loop will call it once for each link in the work. Best to do reindexing link order at the end once work_order is sorted.